lua/entities/gmod_wire_expression2/core/custom/constraintcore.lua

```diff
@@ -1,14 +1,97 @@
+local IsValid = IsValid
+local math_min = math.min
+local math_max = math.max
+local table_insert = table.insert
+
+local cvFlags = {FCVAR_ARCHIVE}
+local maxWeld = CreateConVar( "wire_expression2_max_constraints_weld", "0", cvFlags, 0 )
+local maxRope = CreateConVar( "wire_expression2_max_constraints_rope", "0", cvFlags, 0 )
+local maxAxis = CreateConVar( "wire_expression2_max_constraints_axis", "0", cvFlags, 0 )
+local maxTotal = CreateConVar( "wire_expression2_max_constraints_total", "0", cvFlags, 0 )
+local maxSlider = CreateConVar( "wire_expression2_max_constraints_slider", "0", cvFlags, 0 )
+local maxElastic = CreateConVar( "wire_expression2_max_constraints_elastic", "0", cvFlags, 0 )
+local maxNocollide = CreateConVar( "wire_expression2_max_constraints_nocollide", "0", cvFlags, 0 )
+local maxHydraulic = CreateConVar( "wire_expression2_max_constraints_hydraulic", "0", cvFlags, 0 )
+local maxPerEntity = CreateConVar( "wire_expression2_max_consttraints_per_entity", "0", cvFlags, 0 )
+local maxBallsocket = CreateConVar( "wire_expression2_max_constraints_ballsocket", "0", cvFlags, 0 )
+local maxAdvBallsocket = CreateConVar( "wire_expression2_max_constraints_ballsocket_adv", "0", cvFlags, 0)
+
+local edictCutOff = CreateConVar( "wire_expression2_constraints_edict_cutoff", "0", cvFlags, "At what edict count will E2s be prevented from creating new rope-like constraints (0 turns the check off)", 0, 8192 )
+local shouldCleanup = CreateConVar( "Wire_expression2_constraints_cleanup", "0", cvFlags, "Whether or not Constraint Core should remove all constraints made by an E2 when it's deleted", 0, 1 )
+
+local playerCounts = WireLib.RegisterPlayerTable()
+
+-- Returns the table being used to keep track of counts
+local function getCountHolder(self)
+	local ply = self.player
+	return IsValid( ply ) and playerCounts[ply] or self.data
+end
+
+local function clearCreatedConstraints(self)
+	if not shouldCleanup:GetBool() then return end
+
+	local data = getCountHolder( self )
+	for _, constraint in ipairs( data.allConstraints ) do
+		if constraint:IsValid() then
+			constraint:Remove()
+		end
+	end
+
+	data.allConstraints = {}
+end
+
+local function setupCounts(holder)
+	holder.allConstraints = holder.allConstraints or {}
+	holder.constraintCounts = holder.cosntraintCounts or {}
+	holder.entityConstraints = holder.entityConstraints or {}
+	holder.totalConstraints = holder.totalConstraints or 0
+end
+
 registerCallback("construct", function(self)
+	-- Set up counts on both the E2 and the player in case the player is, or becomes invalid
+	setupCounts( self.data )
+
+	local ply = self.player
+	if IsValid( ply ) then
+		playerCounts[ply] = playerCounts[ply] or {}
+		setupCounts( playerCounts[ply] )
+	end
+
 	self.data.constraintUndos = true
 end)
 
+
+registerCallback("destruct", function(self)
+	clearCreatedConstraints(self)
+end)
+
+hook.Add("PlayerDisconnected", "Wire_Expression2_ConstraintsCleanup", function(ply)
+	playerCounts[ply] = nil
```

Comment:
Redundant because `WireLib.RegisterPlayerTable()` returns a table that will automatically have its keys set to nil as players disconnect